### PR TITLE
§6.1 (phase 6): migrate MATERIALIZATIONS clause onto the lexer/cursor

### PR DESCRIPTION
**Pull request by anentropic**
## Summary

Phase 6 of the §6.1 incremental lexer + cursor migration (Task #26). Parses each `MATERIALIZATIONS` entry — the name, the `AS`, the parenthesized sub-body, and the depth-0 `TABLE` / `DIMENSIONS` / `METRICS` keyword scan — on the shared `Cursor`/lexer instead of the ad-hoc byte scanners (`split_first_token`, `extract_paren_content`, and the local `find_sub_keyword_positions`).

Keyword, `AS`-boundary, and paren detection are now quote- and depth-aware **by construction**:

- **name is the first token** — a double-quoted name with interior whitespace (`"my mat"`) stays whole instead of splitting mid-quote the way the old first-whitespace scan did;
- **`ASx` is a single ident token**, never the `AS` keyword (no `is_ident_continuation` boundary probe needed);
- a **`)` inside a quoted identifier / string** cannot close the sub-body early (PA-6);
- a **bare dimension/metric named like a sub-keyword** (e.g. `metrics`) nested inside a `DIMENSIONS`/`METRICS` list (depth > 0) no longer splits the sub-body.

## Changes

- **`cursor.rs`**: add `find_all_kw_depth0(&[&str]) -> Vec<Token>` — the depth-0 tiling variant of `find_kw_depth0` that collects every matching keyword token in source order. Replaces the local quote+depth `find_sub_keyword_positions` scan. Unit-tested for the depth-0 + quote-aware guarantee.
- **`materializations.rs`**: rewritten onto the cursor. Entry comma-splitting and per-list name splitting stay on the shared `split_at_depth0_commas`.
- **`scan.rs`**: retire the now-dead `split_first_token` and `extract_paren_content` (both were used only by this clause). `extract_paren_prefix` stays — the trailing-annotation parser still needs its consumed-byte count.
- **`mod.rs`**: new regression tests — quote-aware name with a space, leading-symbol missing-name, unquoted nested keyword-like name.

## Behaviour preservation

All existing `MATERIALIZATIONS` unit tests and sqllogictest coverage are unchanged. The parser-level proptests don't generate `MATERIALIZATIONS`, so the contract is the unit tests + sqllogictest, which all pass. The quote/depth-awareness wins are structural (the lexer makes them free). The only behaviour *changes* are on previously-malformed input (quoted-name-with-space now accepted; leading-symbol / nested-keyword edge cases get cleaner handling), each covered by a new regression test.

## Local gate (all green)

- `cargo fmt --check` ✅
- `cargo clippy -- -D warnings` ✅
- `cargo test` ✅ (1186 unit tests + all integration/proptests)
- `make debug` (extension build) ✅
- `make test_debug` (sqllogictest) ✅ — 71 tests run, 0 failed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01JCukus8YB1Qf853dXguuia)_

### CODE REVIEW

File: src/body_parser/materializations.rs [hunk 1]
```diff
@@ -29,90 +41,92 @@ fn parse_single_materialization_entry(
...
-    }
+    let mut cur = Cursor::new(entry, entry_offset);
 
-    // Extract name before "AS"
-    let (name, rest) = split_first_token(entry);
-    if name.is_empty() {
+    // The name is the first token, which must be an identifier (bare or
+    // quoted) — not punctuation. A quoted name keeps its quotes and may contain
+    // whitespace (`"my mat"`), since it is one token now rather than a
+    // first-whitespace split.
```
[the remark this answers is not in the document]
Comment (anentropic) [reply src/body_parser/materializations.rs]: Good catch — reworded (0d30ebb). I kept `peek_is_value()` rather than tightening to identifiers-only, because that preserves the retired `split_first_token` scan's behaviour: it also took a leading `'string'` verbatim as the name. So only a leading *symbol* is rejected as "missing name"; the comment now states that acceptance criterion (bare/quoted identifier or string literal) accurately.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/body_parser/cursor.rs [hunk 2]
```diff
@@ -282,6 +304,21 @@ mod tests {
...
+        // TABLE / DIMENSIONS / METRICS at depth 0 are found in order; a
+        // keyword-like *name* nested inside a (...) list (`METRICS` as a dim
+        // name at depth 1) or hidden inside a quoted ident is inert.
+        let c = Cursor::new(
+            "TABLE t, DIMENSIONS (metrics, \"TABLE\"), METRICS (total)",
+            0,
+        );
+        let kws = c.find_all_kw_depth0(&["TABLE", "DIMENSIONS", "METRICS"]);
+        let texts: Vec<&str> = kws.iter().map(|&t| c.text(t)).collect();
+        // The depth-0 `metrics` name and the quoted `"TABLE"` do not appear.
```
[the remark this answers is not in the document]
Comment (anentropic) [reply src/body_parser/cursor.rs]: Fixed in 0d30ebb — the nested `metrics` is at depth 1 inside `DIMENSIONS (...)`; comment now says "nested (depth-1)".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/body_parser/materializations.rs [hunk 3]
```diff
@@ -29,90 +41,95 @@ fn parse_single_materialization_entry(
...
+            "Expected materialization name in MATERIALIZATIONS entry.".to_string()
+        };
         return Err(ParseError {
-            message: "Expected materialization name in MATERIALIZATIONS entry.".to_string(),
+            message,
             position: Some(entry_offset),
         });
     }
-    let rest = rest.trim();
+    let name_tok = cur.bump().expect("peek_is_value guaranteed a token");
```
Comment (anentropic) [inline src/body_parser/materializations.rs:65]: Optional consistency note (not a regression — this matches old behaviour). `peek_is_value()` is true for an `Unterminated` token, so an unterminated-quote name like `"unclosed AS (TABLE t)` lexes as a single token spanning the whole entry, becomes `name`, and then fails at the AS check with:

> Expected 'AS' after materialization name `"unclosed AS (TABLE t)` in MATERIALIZATIONS clause.

— embedding the entire entry as the "name". The sibling clauses reject this cleanly up front: TABLES checks the token's `Unterminated { ident: true }` kind, and entries/metrics call `unterminated_quote_error(entry)`. Since the lexer already hands you the token kind here, a one-liner after `let name_tok = ...` would bring MATERIALIZATIONS in line and give an "Unterminated quoted identifier in materialization name" message instead:

```rust
if matches!(name_tok.kind, TokenKind::Unterminated { .. }) {
    return Err(cur.err(name_tok.start, format!(
        "Unterminated quoted identifier in materialization name in MATERIALIZATIONS clause."
    )));
}
```

Low priority — MATERIALIZATIONS never had this guard, so it's out of scope for a behaviour-preserving phase; just flagging it since you're consolidating this parser and every other clause already handles it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by anentropic — COMMENTED** [review]
## §6.1 (phase 6): migrate MATERIALIZATIONS clause onto the lexer/cursor

Clean, faithful migration — and notably free of the boundary-semantics divergence that phase 5 had. I traced it against the pre-PR `split_first_token` / `extract_paren_content` / `find_sub_keyword_positions` logic:

- **`find_all_kw_depth0` tiles identically** to the old `find_sub_keyword_positions`: it collects every `TABLE`/`DIMENSIONS`/`METRICS` at bracket-depth 0 in source order, no keyword preference, and each clause's content is sliced from `kw_tok.end` to the next keyword's start — byte-for-byte the same tiling. Depth-0 + quote-awareness are equivalent (a `(` inside a quoted/string token isn't a `Symbol`, so it doesn't nest; a keyword-like name at depth ≥1 is inert), and the token match reproduces the old word-boundary rule (`TABLES` ≠ `TABLE`, `metrics` nested in a dim list is ignored). The new unit test pins exactly this.
- **AS handling** — `is_kw("AS")` matches the old `AS`-with-boundary probe (`AS(` legal, `ASx` not), and the error caret `cur.err(name_tok.end)` reproduces the old `entry_offset + name.len()`.
- **`take_parens`** for the sub-body and each DIMENSIONS/METRICS list matches `extract_paren_content` (quote-aware close), and `extract_paren_list`'s bare-content tolerance is preserved.
- The behaviour *changes* are all on previously-malformed input (quoted-name-with-space now accepted; leading-symbol → clean "missing name"), each covered by a new regression test, and clearly disclosed. Helper deletions (`split_first_token`, `extract_paren_content`) are correctly scoped to this clause's last callers; `extract_paren_prefix` retained for the annotation parser.

One optional consistency note inline: an unterminated-quote name currently falls through to a confusing "Expected 'AS' after name '<whole entry>'" error, where every sibling clause rejects it cleanly — cheap to align now that the lexer hands you the `Unterminated` token kind. Not a regression, not blocking.

Local gate is green (1186 tests + sqllogictest 71/0). Approving in spirit; leaving as **Comment** since I can't run the gate here. This has been a consistently well-executed migration series.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/body_parser/materializations.rs (hunk 3 above)
Comment (anentropic) [reply src/body_parser/materializations.rs:65]: Done in 1f14d24. Added the guard right after `let name_tok = ...`, matching the sibling clauses: it keys off the `Unterminated { ident }` kind and uses the same two nouns those clauses produce — `Unterminated quoted identifier` (`ident: true`) vs `Unterminated string literal` (`ident: false`) — rather than a single message, and phrases it `"{noun} in materialization entry '{entry}'."` to mirror metrics' `"{noun} in metric entry '{entry}'."`. Regression test covers both orphan-quote cases. Full gate green (1186 unit + sqllogictest 71/0).

---
_Generated by [Claude Code](https://claude.ai/code)_
